rust/crd/src/lib.rs

```diff
@@ -804,13 +804,22 @@ impl HdfsCluster {
         Ok(result)
     }
 
-    pub fn is_upgrading(&self) -> bool {
-        self.status
-            .as_ref()
-            .and_then(|status| status.deployed_product_version.as_deref())
-            .map_or(false, |deployed_version| {
-                deployed_version != self.spec.image.product_version()
-            })
+    pub fn upgrade_state(&self) -> Option<UpgradeState> {
+        let status = self.status.as_ref()?;
+        let requested_version = self.spec.image.product_version();
+
+        if requested_version != status.deployed_product_version.as_deref()? {
+            // If we're requesting a different version than what is deployed, assume that we're upgrading.
```

Comment:
**idea**: If HDFS respects semver, we could use the `semver` to determine if it is an upgrade or downgrade.

Reply:
https://hadoop.apache.org/versioning.html looks like the same format as semver, but with slightly different bump criteria.

Either way, the question IMO is closer to "does this change actually _help_ anyone, and is that worth the complexity increase?", and the only improvement I really see in this case would be adding a guardrail to explicitly forbid the operation. But even that could be done simpler by a rule such as "if upgrade_target_product_version is not null then _only_ allow either it or deployed_product_version".

Reply:
So basically in the current code, any change is considered an upgrade (even if the requested version is lower)?

If so, maybe the comment should be less precise?

```suggestion
            // If we're requesting a different version than what is deployed, assume that we're upgrading/downgrading.
```